parser/util/Environment.js

```diff
@@ -0,0 +1,35 @@
+import { LinkedList } from './LinkedList';
+
+/**
+ * A special key-value pair linked list that can lookup values of a certain key.
+ */
+export class Environment extends LinkedList {
+  constructor (head, tail) {
+    super(head, tail);
+    let hed = this.head;
+    this.head = (value) => value === undefined ? hed() : hed([hed()[0], value]);
+  }
+  /**
+   * Creates a new environment with an associated key-value pair.
+   * @param {*} key   the searchable identifier that maps to the value
+   * @param {*} value the value associated with the key
+   * @return a new Environment with the appended key-value pair
+   */
+  push (key, value) {
```

Comment:
1. What will happen if I do `env.push('abc')`? Will it just create a variable with an `undefined` value? Is that the intended usage?
2. Also, what will happen if I do `env.push()`?

Reply:
Noted. Undefined value should be allowed, but undefined key should not (should raise exception). Will do add arity checking in the next commit :)

Reply:
Changed my mind, `undefined` value should not be allowed as well, otherwise the `Environment.lookup` method returning `undefined` will make no sense.